### PR TITLE
Feature/add action plan prefix

### PR DESCRIPTION
**Pull request by torresga**
**Description:**

This PR closes _#23_ . It adds an input that adds a prefix to each item in the action plan.

Before screenshot: 
Screen Shot 2021-06-22 at 4 19 55 PM

After video:

https://user-images.githubusercontent.com/6892410/122994011-f92c5a80-d375-11eb-84ee-d07d6049f10b.mov

I tested this manually in Chrome, Firefox, and Safari.

I will abide by the [code of conduct](https://github.com/fastruby/points/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: app/assets/javascripts/action_plan.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+document.addEventListener("turbolinks:load", function() {
+  $("input[name='action-plan_prefix']").keyup(function(event) {
+    let prefix = event.target.value;
```
Comment (arielj) [inline app/assets/javascripts/action_plan.js:3]: this can be `const prefix` since it's not reasigned

File: app/assets/javascripts/action_plan.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+    $("h3.action-plan_heading").each(function(index) {
+      $(this).children("span").remove();
+      $(this).prepend(`<span>${prefix}${prefix.trim() == "" ? "" : index + 1} </span>`);
```
Comment (arielj) [inline app/assets/javascripts/action_plan.js:7]: you can move the `prefix.trim()` call outside the loop so you don't need to trim the prefix for each iteration

File: app/assets/javascripts/action_plan.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+document.addEventListener("turbolinks:load", function() {
+  $("input[name='action-plan_prefix']").keyup(function(event) {
+    let prefix = event.target.value;
+
+    $("h3.action-plan_heading").each(function(index) {
+      $(this).children("span").remove();
```
Comment (arielj) [inline app/assets/javascripts/action_plan.js:6]: what do you think about always having the `span` element there in the html view and set the innerText value instead of removing/prepending it? it should be more performant

you can do something like

```
$('h3.action-plan_heading > span').each(function(span) {
  span.innerText = `${prefix}${prefix.trim() == "" ? "" : index + 1}`;
})
```

**Review by arielj — COMMENTED** [review]
this is great, I really wanted this feature in the last 2 roadmaps xD

just added some refactor suggestions and it would be great if you can add some system test too

File: app/assets/javascripts/action_plan.js (hunk 3 above)
Comment (torresga) [reply app/assets/javascripts/action_plan.js:6]: @arielj That is okay by me, especially if it is more performant than my version. I will make that change.

**Comment by arielj** [comment]
hey @torresga, I see the comments were marked as resolved but I see no changes in the code, just checking in

**Comment by torresga** [comment]
> hey @torresga, I see the comments were marked as resolved but I see no changes in the code, just checking in

Oops, sorry, I just hadn't pushed my changes yet. I am working on adding some tests now.

**Comment by buys-fran** [comment]
> > hey @torresga, I see the comments were marked as resolved but I see no changes in the code, just checking in
> 
> Oops, sorry, I just hadn't pushed my changes yet. I am working on adding some tests now.

@torresga will you let me know when you would like me to review or QA?

**Comment by torresga** [comment]
Hi @buys-fran , this is ready for review!

File: app/views/action_plans/show.html.erb [hunk 4]
```diff
@@ -3,8 +3,14 @@
 
   <div id="action-plan">
     <h2>Action Plan</h2>
+
+    <label for="action-plan_prefix">
+      Enter Prefix
+      <input name="action-plan_prefix" type="text">
```
Comment (buys-fran) [inline app/views/action_plans/show.html.erb:9]: I think the label's `for` should match the input's `id`
```suggestion
      <input id="action-plan_prefix" name="action-plan_prefix" type="text">
```

File: app/assets/javascripts/action_plan.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+document.addEventListener("turbolinks:load", function() {
+  $("input[name='action-plan_prefix']").keyup(function(event) {
```
Comment (buys-fran) [inline app/assets/javascripts/action_plan.js:2]: If you give the input an id, then you can use the id over here
```suggestion
  $("#action-plan_prefix").keyup(function(event) {
```

File: app/assets/javascripts/action_plan.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    const prefix = event.target.value.trim();
+
+    $("h3.action-plan_heading > span").each(function(index) {
+      $(this).text(`${prefix}${prefix == "" ? "" : index + 1 } `);
```
Comment (buys-fran) [inline app/assets/javascripts/action_plan.js:6]: I noticed that it is using `==` instead of `===` would it need to coerce the prefix to ""?

File: app/views/action_plans/show.html.erb (hunk 4 above)
Comment (buys-fran) [reply app/views/action_plans/show.html.erb:9]: I wonder if you would also prefer using `-` instead of `_` for the id?

File: app/assets/javascripts/action_plan.js (hunk 6 above)
Comment (buys-fran) [reply app/assets/javascripts/action_plan.js:6]: I wonder if you should also add the (dot) when you append the index?
```suggestion
      const suffix = prefix ? "" : `.${index+1}`
      $(this).text(`${prefix}${suffix } `);
```

**Review by buys-fran — COMMENTED** [review]
Hey, it looks good. I left a few comments, let me know what you think.

File: app/views/action_plans/show.html.erb (hunk 4 above)
Comment (torresga) [reply app/views/action_plans/show.html.erb:9]: @buys-fran I agree with that. I will update the name.

File: app/assets/javascripts/action_plan.js (hunk 6 above)
Comment (torresga) [reply app/assets/javascripts/action_plan.js]: > I wonder if you should also add the (dot) when you append the index?

@buys-fran I was wondering this as well, while looking at https://github.com/fastruby/points/issues/23 I assumed that the expected input would be a prefix that looked like `3.2.`. I am ok with either. It's probably a better user experience to add the dot, what do you think?

Comment (buys-fran) [reply app/assets/javascripts/action_plan.js]: I agree, it does seem more user friendly :-)

**Review by arielj — APPROVED** [review]
looks good!

**Review by buys-fran — COMMENTED** [review]
I tested it in chrome. I noticed one very small tweak you could make.
1. Add a prefix
2. Remove the prefix (by pressing backspace)
3. Notice the . (dot) is still there.

I think as a user I would expect the no . (dot) after removing the entire prefix.

image

**Comment by buys-fran** [comment]
> I tested it in chrome. I noticed one very small tweak you could make.
> 
> 1. Add a prefix
> 2. Remove the prefix (by pressing backspace)
> 3. Notice the . (dot) is still there.
> 
> I think as a user I would expect the no . (dot) after removing the entire prefix.
> 
> image

If you think this adds unnecessary complexity, I am okay with proceeding without fixing this edge case. You can remove the dot by refreshing the page.

**Comment by torresga** [comment]
@buys-fran No, I think this is a good edge case. I will update and add tests for this.

**Comment by buys-fran** [comment]
I tested locally and on staging.
Works great 👍

You can merge it, or let me know if you would like me to merge it.

**Comment by torresga** [comment]
@buys-fran Yay! Can you merge it for me please? 😸
